### PR TITLE
Add source map support

### PR DESCRIPTION
**Pull request by TylerHorth**
Adds source map support to the sprockets processor. 

@bouk

### CODE REVIEW

File: lib/sprockets/commoner/railtie.rb [hunk 1]
```diff
@@ -3,7 +3,8 @@ module Commoner
     class Railtie < ::Rails::Railtie
       initializer 'sprockets-commoner' do
         # We need to disable debugging because otherwise Rails will include each file individually, while we need everything to be bundled up together into a single file.
-        config.assets.debug = false
+        # ^ Not applicable to sprockets 4+
+        config.assets.debug = false if Gem::Version.new(Sprockets::VERSION) < Gem::Version.new('4')
```
Comment (TylerHorth) [inline lib/sprockets/commoner/railtie.rb]: Debug mode is required for source maps in sprockets 4.

File: lib/sprockets/commoner.rb [hunk 2]
```diff
@@ -5,6 +5,9 @@
 
 module Sprockets
   module Commoner
+    def self.sprockets4?
+      Gem::Version.new(Sprockets::VERSION) >= Gem::Version.new('4')
```
Comment (bouk) [inline lib/sprockets/commoner.rb]: Maybe cache this value

File: lib/sprockets/commoner/processor.rb [hunk 3]
```diff
@@ -197,14 +205,14 @@ def commoner_options(input)
...
         def options(input)
-          # TODO(bouk): Fix sourcemaps. Sourcemaps are only available in Sprockets v4
           {
             'ast' => false,
             'babelrc' => !match_any?(self.babel_exclude, input[:filename]),
             'filename' => input[:filename],
             'filenameRelative' => PathUtils.split_subpath(input[:load_path], input[:filename]),
             'moduleRoot' => nil,
             'sourceRoot' => @env.root,
+            'sourceMaps' => Commoner.sprockets4?
```
Comment (bouk) [inline lib/sprockets/commoner/processor.rb]: Please keep the trailing comma

File: lib/sprockets/commoner/processor.rb [hunk 4]
```diff
@@ -104,10 +104,18 @@ def call(input)
           @dependencies << "commoner-environment-variable:#{env}"
         end
 
+        map = nil
+        if Commoner.sprockets4?
+          map = result['map']
+          map = Sprockets::SourceMapUtils.decode_vlq_mappings(map['mappings'], sources: map['sources'], names: map['names'])
+          map = Sprockets::SourceMapUtils.combine_source_maps(input[:metadata][:map], map)
+        end
```
Comment (bouk) [inline lib/sprockets/commoner/processor.rb]: Could you put this in a separate method?

File: lib/sprockets/commoner/processor.rb [hunk 5]
```diff
@@ -104,10 +104,18 @@ def call(input)
...
+
         {
           data: result['code'],
           dependencies: @dependencies,
           required: Set.new(@required),
+          map: map,
```
Comment (bouk) [inline lib/sprockets/commoner/processor.rb]: I believe `bundle.rb` also needs to properly update the sourcemap after the header, prelude, and footer are added

File: lib/sprockets/commoner/bundle.rb [hunk 6]
```diff
@@ -78,9 +78,18 @@ def call(input)
         used_helpers = input[:metadata][:commoner_used_helpers].to_a
         header_code = generate_header(used_helpers, global_identifiers)
         {
-          data: "#{PRELUDE}#{header_code}\n#{input[:data]}#{OUTRO}"
+          data: "#{PRELUDE}#{header_code}\n#{input[:data]}#{OUTRO}",
+          map:  shift_map(input[:metadata][:map], PRELUDE.lines.count + header_code.lines.count),
```
Comment (bouk) [inline lib/sprockets/commoner/bundle.rb]: amazing :+1:

**Comment by bouk** [comment]
Thank you Tyler

**Pull request merged**